### ISSUE TITLE
Support `--race` option, when running multiple scripts in one project

### ISSUE THREAD

**Issue by zkochan** [pnpm/pnpm#6792]
pnpm now supports running multiple scripts in one project, using regex (https://pnpm.io/cli/run#running-multiple-scripts).

Hence, we may replace `npm-run-all` in the pnpm repository scripts. But to do so, we also need to implement the `--race` option. With the `--race` option, all tasks will be killed when a task finishes with zero.

**Comment by Bessonov** [comment]
@zkochan What do you think instead of `--parallel` and `--race` allowing to create a graph of tasks with dependencies and regardless of whether it is the same or different module? For example:

```mermaid
stateDiagram-v2
  state "pnpm start" as start
  state "Copy static resources (one-shot)" as copy
  state "Start frontend (long-running)" as frontend
  state "Wait for SSO availability" as sso
  state "Wait for backend availability" as backendrun
  state "Start backend (long-running)" as backend
  start --> copy
  start --> backendrun
  backendrun --> frontend
  start --> sso
  sso --> backend
```

**Comment by zkochan** [comment]
I think there are existing tools that are doing complex script management better than pnpm. I don't know if we want to concentrate too much on this.

**Comment by Shinyaigeek** [comment]
> I think there are existing tools that are doing complex script management better than pnpm. I don't know if we want to concentrate too much on this.

I agree with this. I think more complicated script management should be handled with the tool specialized for that objection.

---

@zkochan I thought I would like to implement to this and considered about how to do that. I want the feedback about the implementation policy from you before the implementation. 

In the current implementation, `pnpm` executes npm scripts with [npm-lifecycle](https://github.com/pnpm/npm-lifecycle) module, which wraps `spawn` and hide process management internally. I think enabling the invoker of `npm-lifecycle` to kill the process generated and hidden by `npm-lifecycle` same as [npm-run-all](https://github.com/mysticatea/npm-run-all/blob/master/lib/run-task.js#L200-L201) does. So I am going to implement it in the following steps.

1. First, I will enable `npm-lifecycle` module to get the `AbortSignal` as the new argument, and make `npm-lifecycle` kill the spawned process on aborted.
2. Second, I wll implement the `--race` flag, and if this flag is turned on, `pnpm` will send abort signal into `npm-lifecycle` module after the fastest task is ended

How do you think about this?

**Comment by Bessonov** [comment]
> I think there are existing tools that are doing complex script management better than pnpm. I don't know if we want to concentrate too much on this.

This is very unfortunate, but I can understand it.

I have re-read the issue:

> With the --race option, all tasks will be killed when a task finishes with zero.

What is the use case for this? I ask because I have never experienced the need.

**Comment by franky47** [comment]
> What is the use case for this? I ask because I have never experienced the need.

End-to-end tests are one example. You run a live server in parallel with a testing tool like Cypress or Playwright, and when it's done testing (no matter the exit code) this will stop the server, which would otherwise keep running indefinitely.

The "no matter the exit code" is important: you wouldn't want a failed test suite leaving your CI running indefinitely. It also follows the behaviour of `Promise.race`, which fulfills when any promise fulfills regardless of the result (resolution or rejection).

I was considering dropping `npm-run-all` (which hasn't had a release in a while and depends on deprecated packages), but I can't at the moment because of this e2e scenario.

**Comment by weyert** [comment]
Can't you use Playwright's webservers config for this? 

That's what I use to run my end-to-end tests in Guthub

**Comment by franky47** [comment]
For Playwright that's what I would do, but I'm using Cypress which doesn't have this feature, and actively recommends against it: https://docs.cypress.io/guides/references/best-practices#Web-Servers

**Comment by weyert** [comment]
I would advise to consider migrating away from Cypress towards Playwright. Cypress already uses Playwright for parts of it :) Trying to contribute to Cypress has left me salty, though.
